request.go

```diff
@@ -95,44 +175,48 @@ func (c *Client) GetRequestStatus(path string) (*RequestStatus, error) {
 	return ret, err
 }
 
+func (c *Client) IsRequestDone(path string) (bool, error) {
+	request, err := c.GetRequestStatus(path)
+	if err != nil {
+		return false, err
+	}
+	switch request.Metadata.Status {
+	case "DONE":
+		return true, nil
+	case "FAILED":
+		return true, NewClientError(
+			RequestFailed,
+			fmt.Sprintf("Request %s failed: %s", request.ID, request.Metadata.Message),
+		)
+	}
+	return false, nil
+}
+
 // WaitTillProvisionedOrCanceled waits for a request to be completed.
 // It returns an error if the request status could not be fetched, the request
 // failed or the given context is canceled.
 func (c *Client) WaitTillProvisionedOrCanceled(ctx context.Context, path string) error {
+	done, err := c.IsRequestDone(path)
+	if err != nil {
+		return err
+	} else if done {
+		return nil
+	}
 	ticker := time.NewTicker(1 * time.Second)
 	defer ticker.Stop()
-	for {
-		select {
-		case <-ctx.Done():
+	for !done {
+		if ctx.Err() != nil {
 			return ctx.Err()
-		default:
 		}
-
-		// The order of a select's cases does not imply a priority for Go.
-		// If the work done on every tick takes longer than the tick interval
-		// and the context is cancelled, both channels can be read.
-		// This is a race condition as the cancelled context might be ignored
-		// indefinitely. To mitigate this behavior, we non-blockingly recheck
-		// the context's cancellation at the beginning of every loop run.
 		select {
-		case <-ctx.Done():
-			return ctx.Err()
 		case <-ticker.C:
-			request, err := c.GetRequestStatus(path)
+			done, err = c.IsRequestDone(path)
```

Comment:
You have a timing issue here. If the context is not canceled at the beginning of the loop, but right after you check it, you will not notice that until the next loop run.
You should either add another `case` to your `select` for  `<-ctx.Done()` or an empty `default:` as your select is currently blocking until the next tick occurs.